File: cpp/Group-Anagrams/Group-Anagrams.cpp
// Solution 1:
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        if (strs.empty()) return res;
        unordered_map<string,int> record;
        for (auto & str : strs)
        {
            string temp = str;
            sort(temp.begin(),temp.end());
            if (record.find(temp) == record.end())
            {
                record[temp] = res.size();
                res.push_back({});
            }
            res[record[temp]].push_back(str);
        }
        return res;
    }
};


// Solution 2:
class Solution {
public:
vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        if (strs.empty()) return res;
        int n = strs.size();
        unordered_map<string,vector<string>> record;
        for (auto str : strs)
        {
            string temp = str;
            sort(temp.begin(),temp.end());
            record[temp].push_back(str);
        }
        for (auto& m:record)
        {
            res.push_back(m.second);  //m是record里边的元素，利用操作符'.'进行索引第2个,下标i是迭代器，用法类似指针，用i->second索引
        }
        // for (auto i = record.begin();i!=record.end();i++)
        // {
        //     res.push_back(i->second);
        // }    
        return res;
    }
};

File: python/Group-Anagrams/Group-Anagrams.py


##### Solution 1:
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = []
        n = len(strs)
        if n==0:
            return res
        dic = {}
        for s in strs:
            t = s
            t = "".join(sorted(t))
            if t not in dic:
                dic[t] = len(res)
                res.append([])
            res[dic[t]].append(s)
        return res
 



### Solution 2:
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = []
        dic = {}
        m ={}
        n = len(strs)
        if n==0: 
            return res
        for s in strs:
            t = s
            t = "".join(sorted(t))
            if t not in dic:
                dic[t] = [s]
            else:
                dic[t].append(s)
        res = list(dic.values())
        return res
